documentation/configuration.md
---
# !!! DO NOT MODIFY !!!
# This file is auto-generated by the gendoc tool based on the source code of the app.
description: This section describes the configuration parameters and their types for your IOTA core node.
keywords:
- IOTA Node 
- Configuration
- JSON
- Customize
- Config
- reference
---


# Configuration

IOTA core node uses a JSON standard format as a config file. If you are unsure about JSON syntax, you can find more information in the [official JSON specs](https://www.json.org).

You can change the path of the config file by using the `-c` or `--config` argument while executing `iota-core` executable.

For example:
```bash
iota-core -c config_example.json
```

You can always get the most up-to-date description of the config parameters by running:

```bash
iota-core -h --full
```

## <a id="app"></a> 1. Application

| Name                      | Description                                            | Type    | Default value |
| ------------------------- | ------------------------------------------------------ | ------- | ------------- |
| checkForUpdates           | Whether to check for updates of the application or not | boolean | true          |
| [shutdown](#app_shutdown) | Configuration for shutdown                             | object  |               |

### <a id="app_shutdown"></a> Shutdown

| Name                     | Description                                                                                            | Type   | Default value |
| ------------------------ | ------------------------------------------------------------------------------------------------------ | ------ | ------------- |
| stopGracePeriod          | The maximum time to wait for background processes to finish during shutdown before terminating the app | string | "5m"          |
| [log](#app_shutdown_log) | Configuration for log                                                                                  | object |               |

### <a id="app_shutdown_log"></a> Log

| Name     | Description                                         | Type    | Default value  |
| -------- | --------------------------------------------------- | ------- | -------------- |
| enabled  | Whether to store self-shutdown events to a log file | boolean | true           |
| filePath | The file path to the self-shutdown log              | string  | "shutdown.log" |

Example:

```json
  {
    "app": {
      "checkForUpdates": true,
      "shutdown": {
        "stopGracePeriod": "5m",
        "log": {
          "enabled": true,
          "filePath": "shutdown.log"
        }
      }
    }
  }
```

## <a id="logger"></a> 2. Logger

| Name        | Description                                                                                                    | Type   | Default value |
| ----------- | -------------------------------------------------------------------------------------------------------------- | ------ | ------------- |
| name        | The optional name of the logger instance. All log messages are prefixed with that name.                        | string | ""            |
| level       | The minimum enabled logging level                                                                              | string | "info"        |
| timeFormat  | Sets the logger's timestamp format. (options: "rfc3339", "rfc3339nano", "datetime", "timeonly", and "iso8601") | string | "rfc3339"     |
| outputPaths | A list of file paths or stdout/stderr to write logging output to                                               | array  | stdout        |

Example:

```json
  {
    "logger": {
      "name": "",
      "level": "info",
      "timeFormat": "rfc3339",
      "outputPaths": [
        "stdout"
      ]
    }
  }
```

## <a id="p2p"></a> 3. Peer to Peer

| Name                                        | Description                                                   | Type   | Default value                                |
| ------------------------------------------- | ------------------------------------------------------------- | ------ | -------------------------------------------- |
| bindMultiAddresses                          | The bind multi addresses for p2p connections                  | array  | /ip4/0.0.0.0/tcp/15600<br/>/ip6/::/tcp/15600 |
| [connectionManager](#p2p_connectionmanager) | Configuration for connectionManager                           | object |                                              |
| externalMultiAddresses                      | External reacheable multi addresses advertised to the network | array  |                                              |
| identityPrivateKey                          | Private key used to derive the node identity (optional)       | string | ""                                           |
| [autopeering](#p2p_autopeering)             | Configuration for autopeering                                 | object |                                              |
| [db](#p2p_db)                               | Configuration for Database                                    | object |                                              |

### <a id="p2p_connectionmanager"></a> ConnectionManager

| Name          | Description                                                                  | Type | Default value |
| ------------- | ---------------------------------------------------------------------------- | ---- | ------------- |
| highWatermark | The threshold up on which connections count truncates to the lower watermark | int  | 10            |
| lowWatermark  | The minimum connections count to hold after the high watermark was reached   | int  | 5             |

### <a id="p2p_autopeering"></a> Autopeering

| Name     | Description                                                              | Type | Default value |
| -------- | ------------------------------------------------------------------------ | ---- | ------------- |
| maxPeers | The max number of autopeer connections. Set to 0 to disable autopeering. | int  | 5             |

### <a id="p2p_db"></a> Database

| Name | Description                  | Type   | Default value      |
| ---- | ---------------------------- | ------ | ------------------ |
| path | The path to the p2p database | string | "testnet/p2pstore" |

Example:

```json
  {
    "p2p": {
      "bindMultiAddresses": [
        "/ip4/0.0.0.0/tcp/15600",
        "/ip6/::/tcp/15600"
      ],
      "connectionManager": {
        "highWatermark": 10,
        "lowWatermark": 5
      },
      "externalMultiAddresses": [],
      "identityPrivateKey": "",
      "autopeering": {
        "maxPeers": 5
      },
      "db": {
        "path": "testnet/p2pstore"
      }
    }
  }
```

## <a id="profiling"></a> 4. Profiling

| Name        | Description                                       | Type    | Default value    |
| ----------- | ------------------------------------------------- | ------- | ---------------- |
| enabled     | Whether the profiling component is enabled        | boolean | false            |
| bindAddress | The bind address on which the profiler listens on | string  | "localhost:6060" |

Example:

```json
  {
    "profiling": {
      "enabled": false,
      "bindAddress": "localhost:6060"
    }
  }
```

## <a id="restapi"></a> 5. RestAPI

| Name                        | Description                                                                                    | Type    | Default value                                                                                                                                                                                                                                                                                                                                                                              |
| --------------------------- | ---------------------------------------------------------------------------------------------- | ------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| bindAddress                 | The bind address on which the REST API listens on                                              | string  | "0.0.0.0:14265"                                                                                                                                                                                                                                                                                                                                                                            |
| publicRoutes                | The HTTP REST routes which can be called without authorization. Wildcards using \* are allowed  | array   | /health<br/>/api/routes<br/>/api/core/v3/info<br/>/api/core/v3/network\*<br/>/api/core/v3/blocks\*<br/>/api/core/v3/transactions\*<br/>/api/core/v3/commitments\*<br/>/api/core/v3/outputs\*<br/>/api/core/v3/accounts\*<br/>/api/core/v3/validators\*<br/>/api/core/v3/rewards\*<br/>/api/core/v3/committee\*<br/>/api/debug/v2/\*<br/>/api/indexer/v2/\*<br/>/api/mqtt/v2<br/>/api/blockissuer/v1/\* |
| protectedRoutes             | The HTTP REST routes which need to be called with authorization. Wildcards using \* are allowed | array   | /api/\*                                                                                                                                                                                                                                                                                                                                                                                     |
| debugRequestLoggerEnabled   | Whether the debug logging for requests should be enabled                                       | boolean | false                                                                                                                                                                                                                                                                                                                                                                                      |
| maxPageSize                 | The maximum number of results per page                                                         | uint    | 100                                                                                                                                                                                                                                                                                                                                                                                        |
| maxCacheSize                | The maximum size of cache for results                                                          | string  | "50MB"                                                                                                                                                                                                                                                                                                                                                                                     |
| [jwtAuth](#restapi_jwtauth) | Configuration for JWT Auth                                                                     | object  |                                                                                                                                                                                                                                                                                                                                                                                            |
| [limits](#restapi_limits)   | Configuration for limits                                                                       | object  |                                                                                                                                                                                                                                                                                                                                                                                            |

### <a id="restapi_jwtauth"></a> JWT Auth

| Name | Description                                                                                                                             | Type   | Default value |
| ---- | --------------------------------------------------------------------------------------------------------------------------------------- | ------ | ------------- |
| salt | Salt used inside the JWT tokens for the REST API. Change this to a different value to invalidate JWT tokens not matching this new value | string | "IOTA"        |

### <a id="restapi_limits"></a> Limits

| Name          | Description                                                               | Type   | Default value |
| ------------- | ------------------------------------------------------------------------- | ------ | ------------- |
| maxBodyLength | The maximum number of characters that the body of an API call may contain | string | "1M"          |
| maxResults    | The maximum number of results that may be returned by an endpoint         | int    | 1000          |

Example:

```json
  {
    "restAPI": {
      "bindAddress": "0.0.0.0:14265",
      "publicRoutes": [
        "/health",
        "/api/routes",
        "/api/core/v3/info",
        "/api/core/v3/network*",
        "/api/core/v3/blocks*",
        "/api/core/v3/transactions*",
        "/api/core/v3/commitments*",
        "/api/core/v3/outputs*",
        "/api/core/v3/accounts*",
        "/api/core/v3/validators*",
        "/api/core/v3/rewards*",
        "/api/core/v3/committee*",
        "/api/debug/v2/*",
        "/api/indexer/v2/*",
        "/api/mqtt/v2",
        "/api/blockissuer/v1/*"
      ],
      "protectedRoutes": [
        "/api/*"
      ],
      "debugRequestLoggerEnabled": false,
      "maxPageSize": 100,
      "maxCacheSize": "50MB",
      "jwtAuth": {
        "salt": "IOTA"
      },
      "limits": {
        "maxBodyLength": "1M",
        "maxResults": 1000
      }
    }
  }
```

## <a id="debugapi"></a> 6. DebugAPI

| Name               | Description                               | Type    | Default value |
| ------------------ | ----------------------------------------- | ------- | ------------- |
| enabled            | Whether the DebugAPI component is enabled | boolean | false         |
| [db](#debugapi_db) | Configuration for Database                | object  |               |

### <a id="debugapi_db"></a> Database

| Name                            | Description                                                | Type   | Default value   |
| ------------------------------- | ---------------------------------------------------------- | ------ | --------------- |
| path                            | The path to the database folder                            | string | "testnet/debug" |
| maxOpenDBs                      | Maximum number of open database instances                  | int    | 2               |
| granularity                     | How many slots should be contained in a single DB instance | int    | 100             |
| [pruning](#debugapi_db_pruning) | Configuration for pruning                                  | object |                 |

### <a id="debugapi_db_pruning"></a> Pruning

| Name      | Description                        | Type | Default value |
| --------- | ---------------------------------- | ---- | ------------- |
| threshold | How many epochs should be retained | uint | 1             |

Example:

```json
  {
    "debugAPI": {
      "enabled": false,
      "db": {
        "path": "testnet/debug",
        "maxOpenDBs": 2,
        "granularity": 100,
        "pruning": {
          "threshold": 1
        }
      }
    }
  }
```

## <a id="metricstracker"></a> 7. MetricsTracker

| Name    | Description                                   | Type    | Default value |
| ------- | --------------------------------------------- | ------- | ------------- |
| enabled | Whether the Metrics Tracker plugin is enabled | boolean | true          |

Example:

```json
  {
    "metricsTracker": {
      "enabled": true
    }
  }
```

## <a id="db"></a> 8. Database

| Name                   | Description                               | Type   | Default value      |
| ---------------------- | ----------------------------------------- | ------ | ------------------ |
| engine                 | The used database engine (rocksdb/mapdb)  | string | "rocksdb"          |
| path                   | The path to the database folder           | string | "testnet/database" |
| maxOpenDBs             | Maximum number of open database instances | int    | 5                  |
| [pruning](#db_pruning) | Configuration for pruning                 | object |                    |

### <a id="db_pruning"></a> Pruning

| Name                     | Description                                  | Type   | Default value |
| ------------------------ | -------------------------------------------- | ------ | ------------- |
| threshold                | How many finalized epochs should be retained | uint   | 30            |
| [size](#db_pruning_size) | Configuration for size                       | object |               |

### <a id="db_pruning_size"></a> Size

| Name                | Description                                                                       | Type    | Default value |
| ------------------- | --------------------------------------------------------------------------------- | ------- | ------------- |
| enabled             | Whether to delete old block data from the database based on maximum database size | boolean | true          |
| targetSize          | Target size of the database                                                       | string  | "30GB"        |
| reductionPercentage | The percentage the database size gets reduced if the target size is reached       | float   | 10.0          |
| cooldownTime        | Cooldown time between two pruning by database size events                         | string  | "5m"          |

Example:

```json
  {
    "db": {
      "engine": "rocksdb",
      "path": "testnet/database",
      "maxOpenDBs": 5,
      "pruning": {
        "threshold": 30,
        "size": {
          "enabled": true,
          "targetSize": "30GB",
          "reductionPercentage": 10,
          "cooldownTime": "5m"
        }
      }
    }
  }
```

## <a id="protocol"></a> 9. Protocol

| Name                             | Description                                                      | Type    | Default value                      |
| -------------------------------- | ---------------------------------------------------------------- | ------- | ---------------------------------- |
| [snapshot](#protocol_snapshot)   | Configuration for snapshot                                       | object  |                                    |
| commitmentCheck                  | Specifies whether commitment and ledger checks should be enabled | boolean | true                               |
| [filter](#protocol_filter)       | Configuration for filter                                         | object  |                                    |
| protocolParametersPath           | The path of the protocol parameters file                         | string  | "testnet/protocol_parameters.json" |
| [baseToken](#protocol_basetoken) | Configuration for baseToken                                      | object  |                                    |

### <a id="protocol_snapshot"></a> Snapshot

| Name  | Description                                                                                | Type   | Default value          |
| ----- | ------------------------------------------------------------------------------------------ | ------ | ---------------------- |
| path  | The path of the snapshot file                                                              | string | "testnet/snapshot.bin" |
| depth | Defines how many slot diffs are stored in the snapshot, starting from the full ledgerstate | int    | 5                      |

### <a id="protocol_filter"></a> Filter

| Name                 | Description                                                                                | Type   | Default value |
| -------------------- | ------------------------------------------------------------------------------------------ | ------ | ------------- |
| maxAllowedClockDrift | The maximum drift our wall clock can have to future blocks being received from the network | string | "5s"          |

### <a id="protocol_basetoken"></a> BaseToken

| Name         | Description                       | Type   | Default value |
| ------------ | --------------------------------- | ------ | ------------- |
| name         | The base token name               | string | "Shimmer"     |
| tickerSymbol | The base token ticker symbol      | string | "SMR"         |
| unit         | The base token unit               | string | "SMR"         |
| subunit      | The base token subunit            | string | "glow"        |
| decimals     | The base token amount of decimals | uint   | 6             |

Example:

```json
  {
    "protocol": {
      "snapshot": {
        "path": "testnet/snapshot.bin",
        "depth": 5
      },
      "commitmentCheck": true,
      "filter": {
        "maxAllowedClockDrift": "5s"
      },
      "protocolParametersPath": "testnet/protocol_parameters.json",
      "baseToken": {
        "name": "Shimmer",
        "tickerSymbol": "SMR",
        "unit": "SMR",
        "subunit": "glow",
        "decimals": 6
      }
    }
  }
```

## <a id="retainer"></a> 10. Retainer

| Name                    | Description                                           | Type    | Default value |
| ----------------------- | ----------------------------------------------------- | ------- | ------------- |
| debugStoreErrorMessages | Whether to store debug error messages in the database | boolean | false         |

Example:

```json
  {
    "retainer": {
      "debugStoreErrorMessages": false
    }
  }
```

## <a id="prometheus"></a> 11. Prometheus

| Name            | Description                                          | Type    | Default value  |
| --------------- | ---------------------------------------------------- | ------- | -------------- |
| enabled         | Whether the Metrics component is enabled             | boolean | true           |
| bindAddress     | Bind address on which the Prometheus exporter server | string  | "0.0.0.0:9311" |
| goMetrics       | Include go metrics                                   | boolean | false          |
| processMetrics  | Include process metrics                              | boolean | false          |
| promhttpMetrics | Include promhttp metrics                             | boolean | false          |

Example:

```json
  {
    "prometheus": {
      "enabled": true,
      "bindAddress": "0.0.0.0:9311",
      "goMetrics": false,
      "processMetrics": false,
      "promhttpMetrics": false
    }
  }
```

## <a id="inx"></a> 12. INX

| Name        | Description                                            | Type    | Default value    |
| ----------- | ------------------------------------------------------ | ------- | ---------------- |
| enabled     | Whether the INX plugin is enabled                      | boolean | false            |
| bindAddress | The bind address on which the INX can be accessed from | string  | "localhost:9029" |

Example:

```json
  {
    "inx": {
      "enabled": false,
      "bindAddress": "localhost:9029"
    }
  }
```



tools/gendoc/configuration_header.md
---
description: This section describes the configuration parameters and their types for your IOTA core node.
keywords:
- IOTA Node 
- Configuration
- JSON
- Customize
- Config
- reference
---


# Configuration

IOTA core node uses a JSON standard format as a config file. If you are unsure about JSON syntax, you can find more information in the [official JSON specs](https://www.json.org).

You can change the path of the config file by using the `-c` or `--config` argument while executing `iota-core` executable.

For example:
```bash
iota-core -c config_example.json
```

You can always get the most up-to-date description of the config parameters by running:

```bash
iota-core -h --full
```



components/restapi/params.go
package restapi

import (
	"github.com/iotaledger/hive.go/app"
)

// ParametersRestAPI contains the definition of the parameters used by REST API.
type ParametersRestAPI struct {
	// the bind address on which the REST API listens on
	BindAddress string `default:"0.0.0.0:14265" usage:"the bind address on which the REST API listens on"`
	// the HTTP REST routes which can be called without authorization. Wildcards using * are allowed
	PublicRoutes []string `usage:"the HTTP REST routes which can be called without authorization. Wildcards using * are allowed"`
	// the HTTP REST routes which need to be called with authorization. Wildcards using * are allowed
	ProtectedRoutes []string `usage:"the HTTP REST routes which need to be called with authorization. Wildcards using * are allowed"`
	// whether the debug logging for requests should be enabled
	DebugRequestLoggerEnabled bool `default:"false" usage:"whether the debug logging for requests should be enabled"`
	// MaxPageSize defines the maximum number of results per page.
	MaxPageSize uint32 `default:"100" usage:"the maximum number of results per page"`
	// MaxCacheSize defines the maximum size of cache for results.
	MaxCacheSize string `default:"50MB" usage:"the maximum size of cache for results"`

	JWTAuth struct {
		// salt used inside the JWT tokens for the REST API. Change this to a different value to invalidate JWT tokens not matching this new value
		Salt string `default:"IOTA" usage:"salt used inside the JWT tokens for the REST API. Change this to a different value to invalidate JWT tokens not matching this new value"`
	} `name:"jwtAuth"`

	Limits struct {
		// the maximum number of characters that the body of an API call may contain
		MaxBodyLength string `default:"1M" usage:"the maximum number of characters that the body of an API call may contain"`
		// the maximum number of results that may be returned by an endpoint
		MaxResults int `default:"1000" usage:"the maximum number of results that may be returned by an endpoint"`
	}
}

var ParamsRestAPI = &ParametersRestAPI{
	PublicRoutes: []string{
		"/health",
		"/api/routes",
		"/api/core/v3/info",
		"/api/core/v3/network*",
		"/api/core/v3/blocks*",
		"/api/core/v3/transactions*",
		"/api/core/v3/commitments*",
		"/api/core/v3/outputs*",
		"/api/core/v3/accounts*",
		"/api/core/v3/validators*",
		"/api/core/v3/rewards*",
		"/api/core/v3/committee*",
		"/api/debug/v2/*",
		"/api/indexer/v2/*",
		"/api/mqtt/v2",
		"/api/blockissuer/v1/*",
	},
	ProtectedRoutes: []string{
		"/api/*",
	},
}

var params = &app.ComponentParams{
	Params: map[string]any{
		"restAPI": ParamsRestAPI,
	},
	Masked: []string{"restAPI.jwtAuth.salt"},
}
